File: 2281.sum-of-total-strength-of-wizards.cpp
/*
 * @lc app=leetcode id=2281 lang=cpp
 *
 * [2281] Sum of Total Strength of Wizards
 */
#include <bits/stdc++.h>

using std::stack;
using std::vector;

// @lc code=start
class Solution {
public:
    // 1. Use a mono-stack to find a subarray [L, R] where the current pivot is 
    //    the smallest element. Then, we need to compute sum of all subarrays 
    //    within [L, R] that include pivot.
    // 2. Use prefix sum and prefix multiple to find sum of all subarrays from pivot to 
    //    L (left_sum), and from pivot to R (right_sum). Then the sum of all subarrays 
    //    is right_sum * (pivot - left + 1) + left_sum * (right - pivot). Finally, we 
    //    multiply this sum by the value of the pivot (smallest) element and add it to 
    //    the result.
    // Detailed explanation: https://leetcode.com/problems/sum-of-total-strength-of-wizards/discuss/2061787
    // T: O(n) amortized
    // S: O(n)
    int totalStrength(vector<int>& strength) {
        const long long n = strength.size(), MOD = 1000000007;
        vector<int> pre_sum_l(n + 1), pre_sum_r(n + 1), pre_mul_l(n + 1), pre_mul_r(n + 1);
        stack<int> st;  // stores indices of strengths ordered ASC (stack top corresponds to weakest)

        for (long long l = 0; l < n; ++l) {
            pre_sum_l[l + 1] = (pre_sum_l[l] + strength[l]) % MOD;
            pre_mul_l[l + 1] = (pre_mul_l[l] + (l + 1) * strength[l]) % MOD;
            
            long long r = n - l - 1;
            pre_sum_r[r] = (pre_sum_r[r + 1] + strength[r]) % MOD;
            pre_mul_r[r] = (pre_mul_r[r + 1] + (n - r) * strength[r]) % MOD;
        }

        long long ans = 0;
        for (long long r = 0; r <= n; ++r) {
            while (!st.empty() && (r == n || strength[st.top()] >= strength[r])) {
                int pivot = st.top(); st.pop();
                long long l = st.empty() ? 0 : st.top() + 1;
                long long l_sum = (MOD + pre_mul_l[pivot + 1] - pre_mul_l[l] - l * (pre_sum_l[pivot + 1] - pre_sum_l[l]) % MOD) % MOD,
                          r_sum = (MOD + pre_mul_r[pivot + 1] - pre_mul_r[r] - ((n - r) * (pre_sum_r[pivot + 1] - pre_sum_r[r]) % MOD)) % MOD;
                long long sum = (r_sum * (pivot - l + 1) + l_sum * (r - pivot)) % MOD;
                ans = (ans + sum * strength[pivot]) % MOD;
            }
            st.push(r);
        }
        return ans;
    }
};
// @lc code=end

File: 2281.sum-of-total-strength-of-wizards.py
from typing import List


class Solution:
    # Prefix sum and mono-stack
    # T: O(n)
    # S: O(n)
    def totalStrength(self, strength: List[int]) -> int:
        MOD = 10 ** 9 + 7
        n = len(strength)

        # Mono-stack for finding subarray minima
        # find next smaller to the right
        nxt_small_r = [n] * n
        stack = []
        for i in range(n):
            while len(stack) > 0 and strength[stack[-1]] > strength[i]:
                nxt_small_r[stack.pop()] = i
            stack.append(i)

        # find next smaller to the left
        nxt_small_l = [-1] * n
        stack = []
        for i in range(n - 1, -1, -1):
            while len(stack) > 0 and strength[stack[-1]] >= strength[i]:
                nxt_small_l[stack.pop()] = i
            stack.append(i)

        # Prefix sum for finding subarray sums
        prefix_sums = [0] * (n + 1)
        for i in range(n):
            prefix_sums[i + 1] = prefix_sums[i] + strength[i]
        prefix_prefix_sums = [0] * (n + 1)  # considers all subarrays
        for i in range(n):
            prefix_prefix_sums[i + 1] = prefix_prefix_sums[i] + prefix_sums[i + 1]

        # Calculate total strength
        ans = 0
        for i in range(n):
            l = nxt_small_l[i]
            r = nxt_small_r[i]
            n_l = i - l
            n_r = r - i
            sum_l = prefix_prefix_sums[i] - prefix_prefix_sums[max(l, 0)]
            sum_r = prefix_prefix_sums[r] - prefix_prefix_sums[i]
            
            # Sum up all subarrays between l and r with strength[i] as the minimum strength.
            # Explanation: 
            #   Suppose this array[l...r]'s indices are: a1, a2, a3, ... aL, aW, b1, b2, b3..., bR, 
            #   with A[aW] be the smallest. 
            #
            #   Now we can pick any index ai from a1 to aL, and any index bj from b1 to bR to form an array 
            #   such that this array is using aW as the min value. 
            #   If we have a prefix sum array, then this particular pick score would be 
            #   A[aW] * (sum[bj] - sum[ai]).
            #
            #   How many combinations of (ai, bj) are there? 
            #   For any particular ai, it can be from sum[b1] to sum[bR], 
            #   so there will be n_l number of sum[b1] to sum[bR], i.e., 
            #   n_l * (sum[b1] + sum[b2] + ... sum[bR]); 
            #   Same idea for bj, we have n_r * (sum[a1] + sum[a2] + ... sum[aL]).
            #
            #   Total score using A[aW]: 
            #   A[aW] * (n_l * (sum[b1] + sum[b2] + ... sum[bR]) - n_r * (sum[a1] + sum[a2] + ... sum[aL])).
            ans += strength[i] * (sum_r * n_l - sum_l * n_r) % MOD

        return ans % MOD
